File: Source/programmers/FWH.vb
'COPYRIGHT EMBEDDED COMPUTERS LLC 2021 - ALL RIGHTS RESERVED
'THIS SOFTWARE IS ONLY FOR USE WITH GENUINE FLASHCATUSB PRODUCTS
'CONTACT EMAIL: support@embeddedcomputers.net
'ANY USE OF THIS CODE MUST ADHERE TO THE LICENSE FILE INCLUDED WITH THIS SDK
'INFO: This object interfaces FCUSB with a Firmware Hub memory device using LPC protocol

Imports FlashcatUSB.FlashMemory
Imports FlashcatUSB.USB

Public Class FWH_Programmer : Implements MemoryDeviceUSB
    Private FCUSB As FCUSB_DEVICE
    Public Property MyFlashDevice As FWH
    Public Property MyFlashStatus As DeviceStatus = DeviceStatus.NotDetected
    Private FLASH_IDENT As FlashDetectResult

    Public Event PrintConsole(message As String) Implements MemoryDeviceUSB.PrintConsole
    Public Event SetProgress(percent As Integer) Implements MemoryDeviceUSB.SetProgress

    Sub New(parent_if As FCUSB_DEVICE)
        FCUSB = parent_if
    End Sub

    Public Function DeviceInit() As Boolean Implements MemoryDeviceUSB.DeviceInit
        Me.MyFlashStatus = DeviceStatus.NotDetected
        Dim ident_data(7) As Byte
        Dim result_data(0) As Byte
        If Not FCUSB.USB_CONTROL_MSG_IN(USBREQ.EXPIO_INIT, result_data, MEM_PROTOCOL.FWH) Then Return False
        If Not (result_data(0) = &H17) Then Return False
        FCUSB.USB_CONTROL_MSG_IN(USBREQ.EXPIO_RDID, ident_data)
        Me.FLASH_IDENT = GetFlashResult(ident_data)
        If Not FLASH_IDENT.Successful Then Return False
        Dim part As UInt32 = (CUInt(FLASH_IDENT.ID1) << 16) Or (FLASH_IDENT.ID2)
        Dim chip_id_str As String = Hex(FLASH_IDENT.MFG).PadLeft(2, "0"c) & Hex(part).PadLeft(8, "0"c)
        RaiseEvent PrintConsole("Mode FWH returned ident code: 0x" & chip_id_str)
        Dim DevList() As Device = FlashDatabase.FindDevices(Me.FLASH_IDENT.MFG, Me.FLASH_IDENT.ID1, 0, MemoryType.FWH_NOR)
        If (DevList.Length = 1) Then
            RaiseEvent PrintConsole(String.Format(RM.GetString("ext_device_detected"), "FWH"))
            MyFlashDevice = CType(DevList(0), FWH)
            MyFlashStatus = DeviceStatus.Supported
            Return True
        Else
            MyFlashDevice = Nothing
            MyFlashStatus = DeviceStatus.NotSupported
            Return False
        End If
    End Function

    Public ReadOnly Property GetDevice As Device Implements MemoryDeviceUSB.GetDevice
        Get
            Return Me.MyFlashDevice
        End Get
    End Property

    Public ReadOnly Property DeviceName As String Implements MemoryDeviceUSB.DeviceName
        Get
            Return MyFlashDevice.NAME
        End Get
    End Property

    Public ReadOnly Property DeviceSize As Long Implements MemoryDeviceUSB.DeviceSize
        Get
            Return MyFlashDevice.FLASH_SIZE
        End Get
    End Property

    Public Function SectorSize(sector As Integer) As Integer Implements MemoryDeviceUSB.SectorSize
        If MyFlashDevice Is Nothing Then Return 0
        Return DirectCast(MyFlashDevice, FWH).SECTOR_SIZE
    End Function

    Public Sub WaitUntilReady() Implements MemoryDeviceUSB.WaitUntilReady
        Utilities.Sleep(100)
    End Sub

    Public Function ReadData(flash_offset As Long, data_count As Integer) As Byte() Implements MemoryDeviceUSB.ReadData
        Dim data_out(data_count - 1) As Byte
        Dim ptr As Integer = 0
        Dim bytes_left As Integer = data_count
        Dim PacketSize As Integer = 2048
        While (bytes_left > 0)
            Dim BufferSize As Integer = bytes_left
            If (BufferSize > PacketSize) Then BufferSize = PacketSize
            Dim data() As Byte = ReadBulk_NOR(CUInt(flash_offset), BufferSize)
            If data Is Nothing Then Return Nothing
            Array.Copy(data, 0, data_out, ptr, BufferSize)
            flash_offset += data.Length
            bytes_left -= data.Length
            ptr += data.Length
        End While
        Return data_out
    End Function

    Public Function WriteData(flash_offset As Long, data_to_write() As Byte, Optional Params As WriteParameters = Nothing) As Boolean Implements MemoryDeviceUSB.WriteData
        Try
            Dim PacketSize As Integer = 4096
            Dim DataToWrite As Integer = data_to_write.Length
            Dim Loops As Integer = CInt(Math.Ceiling(DataToWrite / PacketSize)) 'Calcuates iterations
            For i As Integer = 0 To Loops - 1
                If Params IsNot Nothing Then If Params.AbortOperation Then Return False
                Dim BufferSize As Integer = DataToWrite
                If (BufferSize > PacketSize) Then BufferSize = PacketSize
                Dim data(BufferSize - 1) As Byte
                Array.Copy(data_to_write, (i * PacketSize), data, 0, data.Length)
                Dim ReturnValue As Boolean = WriteBulk_NOR(CUInt(flash_offset), data)
                If (Not ReturnValue) Then Return False
                FCUSB.USB_WaitForComplete()
                flash_offset += data.Length
                DataToWrite -= data.Length
            Next
        Catch ex As Exception
        End Try
        Return True
    End Function

    Public Function EraseDevice() As Boolean Implements MemoryDeviceUSB.EraseDevice
        RaiseEvent SetProgress(0)
        Dim sec_count As Integer = Me.SectorCount()
        Try
            For i = 0 To sec_count - 1
                If Not SectorErase(i) Then Return False
                RaiseEvent SetProgress(CInt((i + 1) / sec_count) * 100)
            Next
        Catch ex As Exception
            Return False
        Finally
            RaiseEvent SetProgress(0)
        End Try
        Return True
    End Function

    Public Function SectorFind(SectorIndex As Integer) As Long Implements MemoryDeviceUSB.SectorFind
        Dim base_addr As UInt32 = 0
        If SectorIndex > 0 Then
            For i As Integer = 0 To SectorIndex - 1
                base_addr += CUInt(Me.SectorSize(i))
            Next
        End If
        Return base_addr
    End Function

    Public Function SectorErase(SectorIndex As Integer) As Boolean Implements MemoryDeviceUSB.SectorErase
        Dim Result As Boolean = False
        Dim Logical_Address As UInt32 = 0
        If (SectorIndex > 0) Then
            For i As Integer = 0 To SectorIndex - 1
                Dim s_size As Integer = SectorSize(i)
                Logical_Address += CUInt(s_size)
            Next
        End If
        Dim erase_setup As UInt32 = (CUInt(MyFlashDevice.ERASE_CMD) << 24) Or Logical_Address
        Result = FCUSB.USB_CONTROL_MSG_OUT(USBREQ.EXPIO_SECTORERASE, Nothing, erase_setup)
        If Not Result Then Return False
        Utilities.Sleep(50)
        Return BlankCheck(Logical_Address)
    End Function

    Public Function SectorCount() As Integer Implements MemoryDeviceUSB.SectorCount
        Return MyFlashDevice.Sector_Count
    End Function

    Public Function SectorWrite(SectorIndex As Integer, data() As Byte, Optional Params As WriteParameters = Nothing) As Boolean Implements MemoryDeviceUSB.SectorWrite
        Dim Addr32 As Long = Me.SectorFind(SectorIndex)
        Return WriteData(Addr32, data, Params)
    End Function

    Private Function BlankCheck(base_addr As UInt32) As Boolean
        Try
            Dim IsBlank As Boolean = False
            Dim Counter As Integer = 0
            Do Until IsBlank
                Utilities.Sleep(10)
                Dim w() As Byte = ReadData(base_addr, 4)
                If w Is Nothing Then Return False
                If w(0) = 255 AndAlso w(1) = 255 AndAlso w(2) = 255 AndAlso w(3) = 255 Then IsBlank = True
                Counter += 1
                If Counter = 50 Then Return False 'Timeout (500 ms)
            Loop
            Return True
        Catch ex As Exception
            Return False
        End Try
    End Function

    Private Function ReadBulk_NOR(address As UInt32, count As Integer) As Byte()
        Try
            Dim read_count As Integer = count
            Dim addr_offset As Boolean = False
            Dim data_out(read_count - 1) As Byte 'Bytes we want to read
            Dim page_size As UInt16 = 512
            If MyFlashDevice IsNot Nothing Then page_size = MyFlashDevice.PAGE_SIZE
            Dim setup_data() As Byte = GetSetupPacket(address, read_count, page_size)
            Dim result As Boolean = FCUSB.USB_SETUP_BULKIN(USBREQ.EXPIO_READDATA, setup_data, data_out, 0)
            If Not result Then Return Nothing
            If addr_offset Then
                Dim new_data(count - 1) As Byte
                Array.Copy(data_out, 1, new_data, 0, new_data.Length)
                data_out = new_data
            Else
                ReDim Preserve data_out(count - 1)
            End If
            Return data_out
        Catch ex As Exception
        End Try
        Return Nothing
    End Function

    Private Function WriteBulk_NOR(address As UInt32, data_out() As Byte) As Boolean
        Try
            Dim setup_data() As Byte = GetSetupPacket(address, data_out.Length, MyFlashDevice.PAGE_SIZE)
            Dim result As Boolean = FCUSB.USB_SETUP_BULKOUT(USBREQ.EXPIO_WRITEDATA, setup_data, data_out, 0)
            Return result
        Catch ex As Exception
        End Try
        Return False
    End Function

    Private Function GetSetupPacket(Address As UInt32, Count As Integer, PageSize As UInt16) As Byte()
        Dim data_in(19) As Byte '18 bytes total
        data_in(0) = CByte(Address And 255)
        data_in(1) = CByte((Address >> 8) And 255)
        data_in(2) = CByte((Address >> 16) And 255)
        data_in(3) = CByte((Address >> 24) And 255)
        data_in(4) = CByte(Count And 255)
        data_in(5) = CByte((Count >> 8) And 255)
        data_in(6) = CByte((Count >> 16) And 255)
        data_in(7) = CByte((Count >> 24) And 255)
        data_in(8) = CByte(PageSize And 255) 'This is how many bytes to increment between operations
        data_in(9) = CByte((PageSize >> 8) And 255)
        Return data_in
    End Function

End Class

File: Source/programmers/FWH.cs
// COPYRIGHT EMBEDDED COMPUTERS LLC 2020 - ALL RIGHTS RESERVED
// THIS SOFTWARE IS ONLY FOR USE WITH GENUINE FLASHCATUSB PRODUCTS
// CONTACT EMAIL: support@embeddedcomputers.net
// ANY USE OF THIS CODE MUST ADHERE TO THE LICENSE FILE INCLUDED WITH THIS SDK
// INFO: This object interfaces FCUSB with a Firmware Hub memory device using LPC protocol

using System;
using System.Linq;
using Microsoft.VisualBasic;

namespace FlashcatUSB {
    public class FWH_Programmer : FlashcatUSB.MemoryDeviceUSB {
        private FlashcatUSB.USB.FCUSB_DEVICE FCUSB;

        public FlashcatUSB.FlashMemory.FWH MyFlashDevice { get; set; }
        public FlashcatUSB.USB.DeviceStatus MyFlashStatus { get; set; } = FlashcatUSB.USB.DeviceStatus.NotDetected;

        private FlashcatUSB.FlashMemory.FlashDetectResult FLASH_IDENT;

        public event PrintConsoleEventHandler PrintConsole;

        public delegate void PrintConsoleEventHandler(string message);

        public event SetProgressEventHandler SetProgress;

        public delegate void SetProgressEventHandler(int percent);

        public FWH_Programmer(FlashcatUSB.USB.FCUSB_DEVICE parent_if) {
            FCUSB = parent_if;
        }

        public bool DeviceInit() {
            MyFlashStatus = FlashcatUSB.USB.DeviceStatus.NotDetected;
            PrintConsole?.Invoke(FlashcatUSB.MainApp.RM.GetString("ext_device_interface") + ": NOR (FWH)");
            var ident_data = new byte[8];
            var result_data = new byte[1];
            if (!FCUSB.USB_CONTROL_MSG_IN(FlashcatUSB.USB.USBREQ.EXPIO_INIT, ref result_data, (uint)FlashcatUSB.MEM_PROTOCOL.FWH))
                return false;
            if (!(result_data[0] == 0x17))
                return false;
            FCUSB.USB_CONTROL_MSG_IN(FlashcatUSB.USB.USBREQ.EXPIO_RDID, ref ident_data);
            FLASH_IDENT = FlashcatUSB.FlashMemory.Tools.GetFlashResult(ident_data);
            if (!FLASH_IDENT.Successful)
                return false;
            uint part = (uint)FLASH_IDENT.ID1 << 16 | (uint)FLASH_IDENT.ID2;
            string chip_id_str = Conversion.Hex(FLASH_IDENT.MFG).PadLeft(2, '0') + Conversion.Hex(part).PadLeft(8, '0');
            PrintConsole?.Invoke("Mode FWH returned ident code: 0x" + chip_id_str);
            var DevList = FlashcatUSB.MainApp.FlashDatabase.FindDevices(FLASH_IDENT.MFG, FLASH_IDENT.ID1, 0, FlashcatUSB.FlashMemory.MemoryType.FWH_NOR);
            if (DevList.Count() == 1) {
                PrintConsole?.Invoke(string.Format(FlashcatUSB.MainApp.RM.GetString("ext_device_detected"), "FWH"));
                MyFlashDevice = (FlashcatUSB.FlashMemory.FWH)DevList[0];
                MyFlashStatus = FlashcatUSB.USB.DeviceStatus.Supported;
                return true;
            } else {
                PrintConsole?.Invoke(FlashcatUSB.MainApp.RM.GetString("unknown_device_email"));
                MyFlashDevice = null;
                MyFlashStatus = FlashcatUSB.USB.DeviceStatus.NotSupported;
                return false;
            }
        }

        public string DeviceName {
            get {
                return MyFlashDevice.NAME;
            }
        }

        public long DeviceSize {
            get {
                return MyFlashDevice.FLASH_SIZE;
            }
        }

        public uint SectorSize(uint sector, FlashcatUSB.FlashMemory.FlashArea area = FlashcatUSB.FlashMemory.FlashArea.Main) {
            if (MyFlashDevice is null)
                return 0U;
            return MyFlashDevice.SECTOR_SIZE;
        }

        public void WaitUntilReady() {
            FlashcatUSB.Utilities.Main.Sleep(100);
        }

        public byte[] ReadData(long flash_offset, long data_count, FlashcatUSB.FlashMemory.FlashArea area = FlashcatUSB.FlashMemory.FlashArea.Main) {
            var data_out = new byte[(int)(data_count - 1L + 1)];
            int ptr = 0;
            int bytes_left = (int)data_count;
            uint PacketSize = 2048U;
            while (bytes_left > 0) {
                int BufferSize = bytes_left;
                if (BufferSize > PacketSize)
                    BufferSize = (int)PacketSize;
                var data = ReadBulk_NOR((uint)flash_offset, (uint)BufferSize);
                if (data is null)
                    return null;
                Array.Copy(data, 0, data_out, ptr, BufferSize);
                flash_offset += data.Length;
                bytes_left -= data.Length;
                ptr += data.Length;
            }

            return data_out;
        }

        public bool WriteData(long flash_offset, byte[] data_to_write, FlashcatUSB.WriteParameters Params = null) {
            try {
                uint PacketSize = 4096U;
                uint BytesWritten = 0U;
                uint DataToWrite = (uint)data_to_write.Length;
                int Loops = (int)Math.Ceiling(DataToWrite / (double)PacketSize); // Calcuates iterations
                for (int i = 0, loopTo = Loops - 1; i <= loopTo; i++) {
                    if (Params is object) {
                        if (Params.AbortOperation)
                            return false;
                    }

                    int BufferSize = (int)DataToWrite;
                    if (BufferSize > PacketSize)
                        BufferSize = (int)PacketSize;
                    var data = new byte[BufferSize];
                    Array.Copy(data_to_write, i * PacketSize, data, 0L, data.Length);
                    bool ReturnValue = WriteBulk_NOR((uint)flash_offset, data);
                    if (!ReturnValue)
                        return false;
                    FCUSB.USB_WaitForComplete();
                    flash_offset += data.Length;
                    DataToWrite = (uint)(DataToWrite - data.Length);
                    BytesWritten = (uint)(BytesWritten + data.Length);
                    if (Params is object && Loops > 1) {
                        uint UpdatedTotal = (uint)(Params.BytesWritten + BytesWritten);
                        float percent = (float)UpdatedTotal / (float)Params.BytesTotal * 100f;
                        if (Params.Status.UpdateSpeed is object) {
                            string speed_str = Strings.Format((object)Math.Round((double)UpdatedTotal / ((double)Params.Timer.ElapsedMilliseconds / 1000d)), "#,###") + " B/s";
                            Params.Status.UpdateSpeed.DynamicInvoke(speed_str);
                        }

                        if (Params.Status.UpdatePercent is object)
                            Params.Status.UpdatePercent.DynamicInvoke((int)percent);
                    }
                }
            } catch (Exception ex) {
            }

            return true;
        }

        public bool EraseDevice() {
            SetProgress?.Invoke(0);
            int sec_count = (int)SectorCount();
            try {
                for (int i = 0, loopTo = sec_count - 1; i <= loopTo; i++) {
                    if (!SectorErase((uint)i))
                        return false;
                    SetProgress?.Invoke((int)((i + 1) / (double)sec_count * 100d));
                }
            } catch (Exception ex) {
                return false;
            } finally {
                SetProgress?.Invoke(0);
            }

            return true;
        }

        public long SectorFind(uint SectorIndex, FlashcatUSB.FlashMemory.FlashArea area = FlashcatUSB.FlashMemory.FlashArea.Main) {
            uint base_addr = 0U;
            if (SectorIndex > 0L) {
                for (uint i = 0U, loopTo = (uint)(SectorIndex - 1L); i <= loopTo; i++)
                    base_addr += SectorSize(i, area);
            }

            return base_addr;
        }

        public bool SectorErase(uint SectorIndex, FlashcatUSB.FlashMemory.FlashArea area = FlashcatUSB.FlashMemory.FlashArea.Main) {
            bool Result = false;
            uint Logical_Address = 0U;
            if (SectorIndex > 0L) {
                for (uint i = 0U, loopTo = (uint)(SectorIndex - 1L); i <= loopTo; i++) {
                    uint s_size = SectorSize(i);
                    Logical_Address += s_size;
                }
            }

            uint erase_setup = (uint)MyFlashDevice.ERASE_CMD << 24 | Logical_Address;
            Result = FCUSB.USB_CONTROL_MSG_OUT(FlashcatUSB.USB.USBREQ.EXPIO_SECTORERASE, null, erase_setup);
            if (!Result)
                return false;
            FlashcatUSB.Utilities.Main.Sleep(50);
            return BlankCheck(Logical_Address);
        }

        public uint SectorCount() {
            return MyFlashDevice.Sector_Count;
        }

        public bool SectorWrite(uint SectorIndex, byte[] data, FlashcatUSB.WriteParameters Params = null) {
            uint Addr32 = (uint)this.SectorFind(SectorIndex, Params.Memory_Area);
            return WriteData(Addr32, data, Params);
        }

        private bool BlankCheck(uint base_addr) {
            try {
                bool IsBlank = false;
                int Counter = 0;
                while (!IsBlank) {
                    FlashcatUSB.Utilities.Main.Sleep(10);
                    var w = this.ReadData((long)base_addr, 4L, FlashcatUSB.FlashMemory.FlashArea.Main);
                    if (w is null)
                        return false;
                    if (w[0] == 255 && w[1] == 255 && w[2] == 255 && w[3] == 255)
                        IsBlank = true;
                    Counter += 1;
                    if (Counter == 50)
                        return false; // Timeout (500 ms)
                }

                return true;
            } catch (Exception ex) {
                return false;
            }
        }

        private byte[] ReadBulk_NOR(uint address, uint count) {
            try {
                uint read_count = count;
                bool addr_offset = false;
                bool count_offset = false;
                var data_out = new byte[(int)(read_count - 1L + 1)]; // Bytes we want to read
                int page_size = 512;
                if (MyFlashDevice is object)
                    page_size = (int)MyFlashDevice.PAGE_SIZE;
                var setup_data = GetSetupPacket(address, read_count, (ushort)page_size);
                bool result = FCUSB.USB_SETUP_BULKIN(FlashcatUSB.USB.USBREQ.EXPIO_READDATA, setup_data, ref data_out, 0U);
                if (!result)
                    return null;
                if (addr_offset) {
                    var new_data = new byte[(int)(count - 1L + 1)];
                    Array.Copy(data_out, 1, new_data, 0, new_data.Length);
                    data_out = new_data;
                } else {
                    Array.Resize(ref data_out, (int)(count - 1L + 1));
                }

                return data_out;
            } catch (Exception ex) {
            }

            return null;
        }

        private bool WriteBulk_NOR(uint address, byte[] data_out) {
            try {
                var setup_data = GetSetupPacket(address, (uint)data_out.Length, (ushort)MyFlashDevice.PAGE_SIZE);
                bool result = FCUSB.USB_SETUP_BULKOUT(FlashcatUSB.USB.USBREQ.EXPIO_WRITEDATA, setup_data, data_out, 0U);
                return result;
            } catch (Exception ex) {
            }

            return false;
        }

        private byte[] GetSetupPacket(uint Address, uint Count, ushort PageSize) {
            var data_in = new byte[20]; // 18 bytes total
            data_in[0] = (byte)(Address & 255L);
            data_in[1] = (byte)(Address >> 8 & 255L);
            data_in[2] = (byte)(Address >> 16 & 255L);
            data_in[3] = (byte)(Address >> 24 & 255L);
            data_in[4] = (byte)(Count & 255L);
            data_in[5] = (byte)(Count >> 8 & 255L);
            data_in[6] = (byte)(Count >> 16 & 255L);
            data_in[7] = (byte)(Count >> 24 & 255L);
            data_in[8] = (byte)(PageSize & 255); // This is how many bytes to increment between operations
            data_in[9] = (byte)(PageSize >> 8 & 255);
            return data_in;
        }
    }
}